src/domain/core/src/services/dependency_graph_service.rs

```diff
@@ -7,75 +7,111 @@
+
+use internal_error::{InternalError, ResultIntoInternal};
 use opendatafabric::DatasetID;
 use thiserror::Error;
 use tokio_stream::Stream;
 
+use crate::{DatasetRepository, GetSummaryOpts};
+
 /////////////////////////////////////////////////////////////////////////////////////////
 
 #[async_trait::async_trait]
 pub trait DependencyGraphService: Sync + Send {
+    /// Forces initialization of graph data, if it wasn't initialized already.
+    /// Ignored if called multiple times
+    async fn eager_initialization(
+        &self,
+        initializer: &DependencyGraphServiceInitializer,
+    ) -> Result<(), InternalError>;
+
     /// Iterates over 1st level of dataset's downstream dependencies
     async fn get_downstream_dependencies(
         &self,
         dataset_id: &DatasetID,
     ) -> Result<DatasetIDStream, GetDownstreamDependenciesError>;
 
-    /// Tracks a dependency between upstream and downstream dataset
-    ///
-    /// TODO: connect to event bus
-    async fn add_dependency(
+    /// Iterates over 1st level of dataset's upstream dependencies
+    async fn get_upstream_dependencies(
         &self,
-        dataset_upstream_id: &DatasetID,
-        dataset_downstream_id: &DatasetID,
-    ) -> Result<(), AddDependencyError>;
-
-    /// Removes tracked dependency between updstream and downstream dataset
-    ///
-    /// TODO: connect to event bus
-    async fn remove_dependency(
-        &self,
-        dataset_upstream_id: &DatasetID,
-        dataset_downstream_id: &DatasetID,
-    ) -> Result<(), RemoveDependencyError>;
+        dataset_id: &DatasetID,
+    ) -> Result<DatasetIDStream, GetUpstreamDependenciesError>;
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[dill::component(pub)]
+pub struct DependencyGraphServiceInitializer {
+    dataset_repo: Arc<dyn DatasetRepository>,
+}
+
+impl DependencyGraphServiceInitializer {
+    pub fn new(dataset_repo: Arc<dyn DatasetRepository>) -> Self {
+        Self { dataset_repo }
+    }
+
+    pub fn browse_dependencies_of_all_datasets(&self) -> DatasetDependenciesIDStream {
+        use tokio_stream::StreamExt;
+
+        Box::pin(async_stream::try_stream! {
+            let mut datasets_stream = self.dataset_repo.get_all_datasets();
+            while let Some(Ok(dataset_handle)) = datasets_stream.next().await {
+                tracing::debug!(dataset=%dataset_handle, "Scanning dataset dependencies");
+
+                let summary = self
+                    .dataset_repo
+                    .get_dataset(&dataset_handle.as_local_ref())
+                    .await
+                    .int_err()?
+                    .get_summary(GetSummaryOpts::default())
+                    .await
+                    .int_err()?;
+
+                let mut upstream_dataset_ids = Vec::new();
+                for transform_input in summary.dependencies.iter() {
+                    if let Some(input_id) = &transform_input.id {
```

Comment:
Strange, why would summary have an ID as optional. We never let create a transform whose aliases are unresolvable.

Reply:
Summary should probably have nothing but an ID - will add this to backwards incompatible changes list.